codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/EventStreamUnmarshallerGenerator.kt

```diff
@@ -0,0 +1,251 @@
+    private val operationShape: OperationShape,
+    private val unionShape: UnionShape,
+) {
+    private val unionSymbol = symbolProvider.toSymbol(unionShape)
+    private val operationErrorSymbol = operationShape.errorSymbol(symbolProvider)
+    private val smithyEventStream = CargoDependency.SmithyEventStream(runtimeConfig)
+    private val codegenScope = arrayOf(
+        "Blob" to RuntimeType("Blob", CargoDependency.SmithyTypes(runtimeConfig), "smithy_types"),
+        "Error" to RuntimeType("Error", smithyEventStream, "smithy_eventstream::error"),
+        "Header" to RuntimeType("Header", smithyEventStream, "smithy_eventstream::frame"),
+        "HeaderValue" to RuntimeType("HeaderValue", smithyEventStream, "smithy_eventstream::frame"),
+        "Inlineables" to RuntimeType.eventStreamInlinables(runtimeConfig),
+        "Message" to RuntimeType("Message", smithyEventStream, "smithy_eventstream::frame"),
+        "SmithyError" to RuntimeType("Error", CargoDependency.SmithyTypes(runtimeConfig), "smithy_types"),
+        "UnmarshallMessage" to RuntimeType("UnmarshallMessage", smithyEventStream, "smithy_eventstream::frame"),
+        "UnmarshalledMessage" to RuntimeType("UnmarshalledMessage", smithyEventStream, "smithy_eventstream::frame"),
+    )
+
+    fun render(): RuntimeType {
+        val unmarshallerType = unionShape.eventStreamUnmarshallerType()
+        return RuntimeType.forInlineFun("${unmarshallerType.name}::new", "event_stream_serde") { inlineWriter ->
+            inlineWriter.renderUnmarshaller(unmarshallerType, unionSymbol)
+        }
+    }
+
+    private fun RustWriter.renderUnmarshaller(unmarshallerType: RuntimeType, unionSymbol: Symbol) {
+        rust(
+            """
+            ##[non_exhaustive]
+            ##[derive(Debug)]
+            pub struct ${unmarshallerType.name};
+
+            impl ${unmarshallerType.name} {
+                pub fn new() -> Self {
+                    ${unmarshallerType.name}
+                }
+            }
+            """
+        )
+
+        rustBlockTemplate(
+            "impl #{UnmarshallMessage} for ${unmarshallerType.name}",
+            *codegenScope
+        ) {
+            rust("type Output = #T;", unionSymbol)
+            rust("type Error = #T;", operationErrorSymbol)
+
+            rustBlockTemplate(
+                """
+                fn unmarshall(
+                    &self,
+                    message: &#{Message}
+                ) -> std::result::Result<#{UnmarshalledMessage}<Self::Output, Self::Error>, #{Error}>
+                """,
+                *codegenScope
+            ) {
+                rustBlockTemplate(
+                    """
+                    let response_headers = #{Inlineables}::parse_response_headers(&message)?;
+                    match response_headers.message_type.as_str()
+                    """,
+                    *codegenScope
+                ) {
+                    rustBlock("\"event\" => ") {
```

Comment:
is this an event stream protocol detail?

Reply:
Yeah, the `:message-type` header will either be `event` or `exception`.